src/metabase/util/analytics.clj

```diff
@@ -0,0 +1,119 @@
+           org.apache.http.impl.conn.PoolingHttpClientConnectionManager))
+
+(defsetting snowplow-url
+  (deferred-tru "The URL of the Snowplow collector to send analytics events to")
+  :default (if config/is-prod?
+             "https://sp.metabase.com"
+             ;; See the iglu-schema-registry repo for instructions on how to run Snowplow Micro locally for development
+             "http://localhost:9095")
+  :visibility :public)
+
+(def ^:private ^Emitter emitter
+  "An instance of a Snowplow emitter"
+  (delay
+    (let [client (-> (HttpClients/custom)
+                     (.setConnectionManager (PoolingHttpClientConnectionManager.))
+                     (.build))
+          builder (-> (ApacheHttpClientAdapter/builder)
+                      (.httpClient client)
+                      (.url (snowplow-url)))
+          adapter (.build ^ApacheHttpClientAdapter$Builder builder)
+          batch-emitter-builder (-> (BatchEmitter/builder)
+                                    (.bufferSize 1)
+                                    (.httpClientAdapter adapter))]
+      (.build ^BatchEmitter$Builder batch-emitter-builder))))
+
+(def ^:private ^Tracker tracker
+  "An instance of a Snowplow tracker"
+  (delay
+   (-> (Tracker$TrackerBuilder. ^Emitter @emitter "sp" "metabase")
+       (.build))))
+
+(defn- subject
+  "Create a Subject object for a given user ID, to be included in analytics events"
+  [user-id]
+  (-> (Subject$SubjectBuilder.)
+      (.userId (str user-id))
+      (.build)))
+
+(defn- context
+  "Common context included in every analytics event"
+  []
+  (new SelfDescribingJson
+       "iglu:com.metabase/instance/jsonschema/1-0-0"
+       {"id"             (public-settings/analytics-uuid),
+        "version"        {"tag" (:tag (public-settings/version))},
+        "token-features" (into {} (for [[token enabled?] (public-settings/token-features)]
+                                    [(name token) enabled?]))}))
+
+(defn- payload
+  "A SelfDescribingJson object containing the provided event data, which can be included as the payload for an
+  analytics event"
+  [schema version event-data]
+  (new SelfDescribingJson
+       (format "iglu:com.metabase/%s/jsonschema/%s" (name schema) version)
+       ;; Make sure keywords are converted to strings
+       (into {} (for [[k v] event-data] [(name k) (if (keyword? v) (name v) v)]))))
+
+(defn- track-schema-event
+  "Send a single analytics event to the Snowplow collector, if tracking is enabled for this MB instance"
+  [schema version user-id event-data]
+  (when (public-settings/anon-tracking-enabled)
+    (try
+      (let [^Unstructured$Builder builder (-> (. Unstructured builder)
+                                              (.eventData (payload schema version event-data))
+                                              (.customContext [(context)]))
+            ^Unstructured$Builder builder' (if user-id
+                                             (.subject builder (subject user-id))
```

Comment:
can you do `(.subject builder nil)`? In that case you could do `(.subject builder (some-> user-id subject))` and then you can do this as part of the `->` form above

Reply:
alternative just write a `(set-subject! builder user-id)` function that handles a `nil` `user-id`